docs/writing_plugins.md

```diff
@@ -0,0 +1,550 @@
+
+The generator functions are invoked by the plugin framework with a parameter of type [`Schema`](#schema).  This object contains the information needed to generate code.  In addition to the [`CodeGeneratorRequest`](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/compiler/plugin.proto) that is standard when working with protoc plugins, the [`Schema`](#schema) object also contains some convenience interfaces that make it a bit easier to work with the various descriptor objects.  
+
+For example, the [`Schema`](#schema) object contains a [`files`](#files) property, which is a list of `DescFile` objects representing the files requested to be generated.  The first thing you will most likely do in your generator function is iterate over this list and issue a call to the [`generateFile`](#generatefile) function for each file object.  This will return a generated file object containing a `print` function which you can then use to "print" to the file.
+
+Each `DescFile` object in the [`files`](#files) property contains all enums and messages needed to begin generating code:
+
+```ts
+// protoc-gen-foo-plugin.js
+
+function generateTs(schema: Schema) {
+   for (const file of schema.files) {
+     const f = schema.generateFile(file.name + "_pb.ts");
+  
+     for (const enumeration of file.enums) {
+        f.print(`// generating enums from ${file.name}`);
+        ...
+     }
+  
+     for (const message of file.messages) {
+        f.print(`// generating messages from ${file.name}`);
+        ...
+     }
+   }
+}
+```
+
+#### Reading Custom Options
+
+Protobuf-ES does not yet provide support for extensions, neither in general as pertaining to proto2 nor with custom options in proto3.  However, in the interim, there are convenience functions for retrieving any custom options specified in your .proto files.  These are provided as a temporary utility until full extension support is implemented.
+
+For available functions, see the [custom options](#custom-options) section under **API**.
+
+
+### Initializing your plugin
+
+Once your generator functions are defined, you then need to initialize your plugin by calling `createEcmaScriptPlugin` with a plugin initialization object.  
+
+```ts
+// protoc-gen-foo-plugin.js
+
+export const protocGenFoo = createEcmaScriptPlugin({
+   name: "protoc-gen-foo",
+   version: "v0.1.0",
+   generateTs,
+});
+```
+
+See the [`PluginInit`](#plugininit) object definition for more information.
+
+### Create an executable file
+
+Then, you need to create an executable file that will execute the plugin returned from the above call:
+
+```js
+// bin/protoc-gen-foo
+
+#!/usr/bin/env node
```

Comment:
The hashbang (#!) needs to be the first line in the file. The comment might throw users off. Can you instead mention in the text that this file should to be put to `bin/protoc-gen-foo` _without_ a file extension ?